File: backend/app/translator/baidu_web.py
import re
from typing import List

import js2py
import requests

from app.translator.baidu import BaiduBaseTranslate

_JS_CODE = """
function a(r, o) {
    for (var t = 0; t < o.length - 2; t += 3) {
        var a = o.charAt(t + 2);
        a = a >= "a" ? a.charCodeAt(0) - 87 : Number(a),
        a = "+" === o.charAt(t + 1) ? r >>> a: r << a,
        r = "+" === o.charAt(t) ? r + a & 4294967295 : r ^ a
    }
    return r
}
var C = null;
var token = function(r, _gtk) {
    var o = r.length;
    o > 30 && (r = "" + r.substr(0, 10) + r.substr(Math.floor(o / 2) - 5, 10) + r.substring(r.length, r.length - 10));
    var t = void 0,
    t = null !== C ? C: (C = _gtk || "") || "";
    for (var e = t.split("."), h = Number(e[0]) || 0, i = Number(e[1]) || 0, d = [], f = 0, g = 0; g < r.length; g++) {
        var m = r.charCodeAt(g);
        128 > m ? d[f++] = m: (2048 > m ? d[f++] = m >> 6 | 192 : (55296 === (64512 & m) && g + 1 < r.length && 56320 === (64512 & r.charCodeAt(g + 1)) ? (m = 65536 + ((1023 & m) << 10) + (1023 & r.charCodeAt(++g)), d[f++] = m >> 18 | 240, d[f++] = m >> 12 & 63 | 128) : d[f++] = m >> 12 | 224, d[f++] = m >> 6 & 63 | 128), d[f++] = 63 & m | 128)
    }
    for (var S = h,
    u = "+-a^+6",
    l = "+-3^+b+-f",
    s = 0; s < d.length; s++) S += d[s],
    S = a(S, u);
    return S = a(S, l),
    S ^= i,
    0 > S && (S = (2147483647 & S) + 2147483648),
    S %= 1e6,
    S.toString() + "." + (S ^ h)
}
"""

_HEADERS = {
    "User-Agent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_2) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/71.0.3578.98 Safari/537.36"
}


class BaiduWebTranslate(BaiduBaseTranslate):
    translator_id = "baidu-web"

    def __init__(self, from_lang: str, to_lang: str) -> None:
        super().__init__(from_lang, to_lang)

        self.session = requests.Session()
        self.token = None
        self.gtk = None
        self.javascript = js2py.eval_js(_JS_CODE)

        # 必须要加载两次保证token是最新的，否则会出现998的错误
        self._load_main_page()
        self._load_main_page()

    def _load_main_page(self):
        url = "https://fanyi.baidu.com"
        res = self.session.get(url, headers=_HEADERS)
        self.token = re.findall(r"token: '(.*?)',", res.text)[0]
        self.gtk = re.findall(r'window.gtk = "(.*?)";', res.text)[0]

    def _inner_translate(self, query: str) -> List[str]:
        url = "https://fanyi.baidu.com/v2transapi"
        sign = self.javascript(query, self.gtk)
        data = {
            "from": self.from_lang,
            "to": self.to_lang,
            "query": query,
            "simple_means_flag": 3,
            "sign": sign,
            "token": self.token,
            "domain": "common",
        }
        res = self.session.post(url=url, data=data, headers=_HEADERS)
        res.raise_for_status()

        json = res.json()
        if "error" in json:
            # 998表示需要重新加载主页获取新的token
            # 1022可能表示超出使用限制
            msg = json["error"]
            raise Exception(f"Baidu translator error: {msg}")
        else:
            result = []
            for data in json["trans_result"]["data"]:
                result.append(data["dst"])
            return result

File: src-vue/src/data/translator/baidu-web.ts
import ky from 'ky';
import { BaiduBaseTranslator } from './baidu';
import { Glossary } from './base';

function a(r: number, o: string) {
  for (let t = 0; t < o.length - 2; t += 3) {
    const a = o.charAt(t + 2);
    let aa = a >= 'a' ? a.charCodeAt(0) - 87 : Number(a);
    aa = '+' === o.charAt(t + 1) ? r >>> aa : r << aa;
    r = '+' === o.charAt(t) ? (r + aa) & 4294967295 : r ^ aa;
  }
  return r;
}
var token = function (r: string, gtk: string) {
  if (r.length > 30) {
    r =
      '' +
      r.substring(0, 10) +
      r.substring(Math.floor(r.length / 2) - 5, 10) +
      r.substring(r.length, r.length - 10);
  }

  const gtkList = gtk.split('.');
  const gtk1 = Number(gtkList[0]) || 0;
  const gtk2 = Number(gtkList[1]) || 0;

  let d = [];
  let j = 0;
  for (let i = 0; i < r.length; i++) {
    var m = r.charCodeAt(i);
    if (128 > m) {
      d[j++] = m;
    } else if (2048 > m) {
      d[j++] = (m >> 6) | 192;
    } else {
      if (
        55296 === (64512 & m) &&
        i + 1 < r.length &&
        56320 === (64512 & r.charCodeAt(i + 1))
      ) {
        m = 65536 + ((1023 & m) << 10) + (1023 & r.charCodeAt(++i));
        d[j++] = (m >> 18) | 240;
        d[j++] = ((m >> 12) & 63) | 128;
      } else {
        d[j++] = (m >> 12) | 224;
        d[j++] = ((m >> 6) & 63) | 128;
      }
      d[j++] = (63 & m) | 128;
    }
  }

  let S = gtk1;
  for (let s = 0; s < d.length; s++) {
    S = a(S + d[s], '+-a^+6');
  }
  S = a(S, '+-3^+b+-f');
  S ^= gtk2;
  0 > S && (S = (2147483647 & S) + 2147483648);
  S %= 1e6;
  return S.toString() + '.' + (S ^ gtk1);
};

export class BaiduWebTranslator extends BaiduBaseTranslator {
  private token = '';
  private gtk = '';

  static async createInstance(
    langSrc: string,
    langDst: string,
    glossary: Glossary
  ) {
    const translator = new this(langSrc, langDst, glossary);
    await translator.loadMainPage();
    await translator.loadMainPage();
    return translator;
  }

  private async loadMainPage() {
    const html = await ky
      .get('https://fanyi.baidu.com', { credentials: 'include' })
      .text();
    this.token = html.match(/token: '(.*?)',/)!![1];
    this.gtk = html.match(/window.gtk = "(.*?)";/)!![1];
  }

  async translateInner(textSrc: string): Promise<string[]> {
    const url = 'https://fanyi.baidu.com/v2transapi';
    const sign = token(textSrc, this.gtk);
    const data = {
      from: this.langSrc,
      to: this.langDst,
      query: textSrc,
      simple_means_flag: 3,
      sign: sign,
      token: this.token,
      domain: 'common',
    };

    const json: any = await ky
      .post(url, { json: data, credentials: 'include' })
      .json();
    if ('error' in json) {
      throw Error(`Baidu translator error ${json.error}: ${json.msg}`);
    } else if ('errno' in json) {
      if (json.errno == 1000) {
        throw Error(
          `Baidu translator error ${json.errno}: ${json.errmsg}，可能是因为输入为空`
        );
      } else {
        throw Error(`Baidu translator error ${json.errno}: ${json.errmsg}`);
      }
    } else {
      return json.trans_result.data.map((item: any) => item.dst);
    }
  }
}
